lib/rex/socket.rb

```diff
@@ -228,7 +226,12 @@ def self.gethostbyname(host)
       host, _ = host.split('%', 2)
     end
 
-    ::Socket.gethostbyname(host)
+    res = ::Addrinfo.getaddrinfo(host, 0, ::Socket::AF_UNSPEC, ::Socket::SOCK_STREAM)
+    res.map! do |address_info|
+      address_info.ip_address
+    end
+
+    res
```

Comment:
Taking a step back, if this file is just a wrapper around Ruby's `::Socket`, and Ruby themselves have marked `Socket::gethostbyname` as deprecated, maybe we need to do the same - and add an additional method for the newer API, and update metasploit-framework to align with this new API

This is potentially more effort, but potentially worth considering the scope of such a change

Reply:
So here, `self.gethostbyname(host)` is only called once to return the byte address here:
```Ruby
def self.resolv_nbo(host)
  self.gethostbyname( Rex::Socket.getaddress(host, true) )[3]
end
```
Before the only information being used was the 4th element of the array:
```
["localhost", [], 30, "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01", "\x7F\x00\x00\x01"]
```
So as our new functionality returns IP addresses, we added this new code to convert the returned value to a byte address.
```Ruby
def self.resolv_nbo(host)
  ip_address = Rex::Socket.getaddress(host)
  IPAddr.new(ip_address).hton
end
```
Previous output:
```
"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01"
```
New output:
```
"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01"
```

Reply:
@cgranleese-r7 If I understand correctly, in the scenario of the host being an ipv4 address the logic is short circuited and an array is returned immediately without resolving the host name:
```
require 2.6.5 :001 > require 'rex/socket'
 => true 
2.6.5 :002 > ::Rex::Socket.methods.grep /host/
 => [:gethostbyname] 
2.6.5 :003 > ::Rex::Socket.gethostbyname("127.0.0.1")
 => ["127.0.0.1", [], 2, "\x7F\x00\x00\x01"] 
```

Otherwise we need to call the real method:
```
2.6.5 :004 > ::Rex::Socket.gethostbyname("google.com")
 => ["2a00:1450:4009:81c::200e", "216.58.211.174"] 
```

In this scenario the result is a very different shape, and any calling code will be broken 👀

Reply:
After discussing with @cgranleese-r7 we might pursue aligning with Ruby's decision to deprecate `gethostbyname` in its entirety.

After looking at metasploit-framework's use of the gethostbyname method, it looks like the common use case is just getting a single address, so an upgrade path might just be something like:

```diff
diff --git a/lib/rex/proto/proxy/socks4a.rb b/lib/rex/proto/proxy/socks4a.rb
index 0c230e4b3d..e3fc73b64b 100644
--- a/lib/rex/proto/proxy/socks4a.rb
+++ b/lib/rex/proto/proxy/socks4a.rb
@@ -132,7 +132,7 @@ class Socks4a
       def resolve( hostname )
         if( not hostname.empty? )
           begin
-            return Rex::Socket.addr_itoa( Rex::Socket.gethostbyname( hostname )[3].unpack( 'N' ).first )
+            return Rex::Socket.getaddress(hostname, false)
           rescue ::SocketError
             return nil
           end
diff --git a/modules/auxiliary/admin/backupexec/dump.rb b/modules/auxiliary/admin/backupexec/dump.rb
index 95ad4c6153..94fe335b88 100644
--- a/modules/auxiliary/admin/backupexec/dump.rb
+++ b/modules/auxiliary/admin/backupexec/dump.rb
@@ -129,7 +129,7 @@ class MetasploitModule < Msf::Auxiliary
       0,
       0,
       1,
-      Rex::Socket.gethostbyname(local_addr)[3],
+      Rex::Socket.resolve_nbo(local_addr, false),
       local_port
     ].pack('NNNNNNNA4N')
diff --git a/modules/auxiliary/dos/windows/smb/ms11_019_electbowser.rb b/modules/auxiliary/dos/windows/smb/ms11_019_electbowser.rb
index 7e1755717d..c5bcd14856 100644
--- a/modules/auxiliary/dos/windows/smb/ms11_019_electbowser.rb
+++ b/modules/auxiliary/dos/windows/smb/ms11_019_electbowser.rb
@@ -56,7 +56,7 @@ class MetasploitModule < Msf::Auxiliary
     @client = Rex::Proto::SMB::Client.new(udp_sock)
     ip = Rex::Socket.source_address(datastore['RHOST'])
-    ip_src = Rex::Socket.gethostbyname(ip)[3]
+    ip_src = Rex::Socket.resolve_nbo(ip, false)
     svc_src = "\x41\x41\x00"   # pre-encoded?
     name_src = Rex::Text.rand_text_alphanumeric(15) # 4+rand(10))
```

There's some internal repositories that make use of `gethostbyname`, so we're on the fence about being good citizens and updating that code too - or leaving that upgrade effort to those teams.
So potential next steps are:

- Leave `::Rex::Socket.gethostbyname` as it is currently, mark it as deprecated, update metasploit-framework to use the new API, and release a future gem that removes it
- Try to update `::Rex::Socket.gethostbyname` to get similar data from `Socket.getaddrinfo`, and add our own deprecation note, and still update metasploit-framework to use the new API. This is probably the most backwards compatible / less impactful change,

Reply:
Let's revert this method entirely, mark it as deprecated, and update metasploit framework to no longer use this method :+1:

Reply:
Sounds good, I'll get that sorted 👍